UNet Segmentation Network + Example (road segmentation)

FabianIsensee opened the pull request:
Everything should work out of the box. If not please let me know.

Downloading the dataset take quite a while. Is there any way to store the images as a .zip archive somewhere (~8GB)? (tiffs are uncompressed and download is slow because they are individual files)

If something needs changing just let me know!

Cheers,

Fabian

FabianIsensee commented:
I just checked my layer activations and noticed that the Deconv Layers look weird. Could you check if that is the case for you as well?
For my own experiments i switched to Upscale2DLayer

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+    g = open("mass_roads_train_target_download.sh", 'w')
+    for img_name in train_data_url:
+        f.write("wget -O data/training/sat_img/%sf "%img_name + train_data_str + img_name + "f" + "\n")
+        g.write("wget -O data/training/map/%s "%img_name + train_target_str + img_name + "\n")
```
Can't you download the files directly, using urlretrieve, similar to the MNIST example: https://github.com/Lasagne/Lasagne/blob/5a009f9/examples/mnist.py#L33-L41 ? The current solution won't work on Windows.
You can assemble a list of source URLs and target file names, create a `multiprocessing.dummy.Pool` (like `multiprocessing.Pool`, but using threads) and call its `map()` method to parallelize the downloads.

f0k replied:
To avoid redownloading, while assembling the list of URLs and target files, check whether the target already exists and skip it. If the list of URLs ends up being empty, you know the dataset is already complete and can avoid downloading.

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+    download_dataset()
+    # the dataset is now downloaded in the background. Once every few seconds we check if the download is done. We do
+    # this by checking whether tha last training image exists
+    while not os.path.isfile("data/training/map/99238675_15.tiff") and not os.path.isfile("data/training/sat_img/99238675_15.tiff"):
```
You can't be 100% sure that the last training image is the last one to be downloaded, though -- if everything happens in parallel, one of the shorter scripts might still end later than the longest script.
Anyway, if you use `map()` as suggested above, you won't need to poll.

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+    np.random.seed()
+    idx = np.arange(data.shape[0])
+    while True:
+        ids = np.random.choice(idx, BATCH_SIZE)
```
You can directly do:
```
ids = np.random.choice(len(data), BATCH_SIZE)
```
`np.random.choice` supports an integer as the first argument.

FabianIsensee replied:
Absolutely. That was a remnant of another experiment of mine (where train and validation images were in the same array and the generator knew which indices were train and which were validation -> np.random.choice(train_indices, BATCH_SIZE) )

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+    # it would not be penalized enough for missclassifying terrain pixels which results in too many false positives)
+    class_weights = np.log(class_frequencies[[1,0]])
+    class_weights = class_weights / np.sum(class_weights) * 2.
+    class_weights = class_weights.astype(np.float32)
```
If you move this above the definition of the loss, you can just do:
```
w_sym = class_weights[seg_sym]
```
Then you don't have to compute and pass the weights to the training and validation function, they'll compute the weights from the targets themselves.

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+
+    # load the prepared data. They have been converted to np arrays because they are much faster to load than single image files.
+    # you will need some ram in order to have everything in memory.
+    # TODO is it worth the effort to refactor everything so that we use memmaps?
```
No need to refactor anything: Just add `mmap_mode='r'` to each `np.load` call. Probably this should be optional, though. If you have enough RAM, loading them before will be faster because then the data is read sequentially, not in random order as requested during training.

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+
+    # if you wish to load pretrained weights you can uncomment this code and modify the file name
+    # if you want, use my pretained weights (got around 96% accuracy and a loss of 0.11 using excessive data
+    # augmentation)
```
Can you elaborate? "using excessive data augmentation: rotation, cropping, flipping, x, y, z)"

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+
+    # there is no need for data augmentation on the validation. However we need patches of the same size which is why
+    # we are using the random crop generator here again
+    validation_generator = random_crop_generator(batch_generator(data_valid, target_valid, BATCH_SIZE), PATCH_SIZE)
```
Can you make this deterministic? Possibly add a `shuffle=False` argument to the `batch_generator` and have it return all the validation data in sequence (or a subset of the data if this is too expensive), and add a `center_crop_generator()` instead of the `random_crop_generator()`. Your current solution is probably still giving you a good idea about how well the network learns, but if we want this to be an educational example, we should show people to use a deterministic validation set :)

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+        for data, target in train_generator:
+            # the output of the net has shape (BATCH_SIZE, N_CLASSES). We therefore need to flatten the segmentation so
+            # that we can match it with the prediction via the crossentropy loss function
+            target_flat = target.flatten()
```
`flatten()` always creates a copy. Prefer `ravel()`. (Again, not a big deal here, but if this is taken to be an educational example...)

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+    '''
+    just a simple batch iterator, no cropping, no rotation, no anything
+    '''
+    np.random.seed()
```
Not really useful here. Everything's running in the main thread, which already has a seeded random generator.

f0k commented on examples/UNet/massachussets_road_segm.py:
```diff
@@ -0,0 +1,340 @@
+    '''
+    yields a random crop of size crop_size
+    '''
+    np.random.seed()
```
This re-seeds the (global) random generator shortly after `batch_generator()` seeded it. I think both calls should be removed.

FabianIsensee replied:
Both calls are unnecessary here, that is true. In this specific example I would agree that the calls could be removed. I would still like to keep the call in batch_generator because some people (like me) like to use multi threaded batch generation (which is necessary when applying for example elastic deformations). In absence of a np.random.seed(None) call each worker would generate exactly the same training data

FabianIsensee commented:
Wow thanks for your suggestions! This is really helpful (and I am still learning myself)!
I will implement your changes tomorrow =)

f0k commented on modelzoo/Unet.py:
```diff
@@ -0,0 +1,62 @@
+    net['dimshuffle'] = DimshuffleLayer(net['segLayer'], (1, 0, 2, 3))
+    net['reshapeSeg'] = ReshapeLayer(net['dimshuffle'], (num_output_classes, -1))
+    net['dimshuffle2'] = DimshuffleLayer(net['reshapeSeg'], (1, 0))
+    net['output'] = NonlinearityLayer(net['dimshuffle2'], nonlinearity=lasagne.nonlinearities.softmax)
```
Maybe it would be nicer to call this `'flat_output'`, and dimshuffle/reshape things back for an `'output'` layer of shape (batchsize, classes, height, width)? You can then use `'flat_output'` for the loss, and people can use `'output'` for getting predictions in a sensible layout.

We should really get https://github.com/Lasagne/Lasagne/issues/626 done to make this less ugly...

FabianIsensee replied:
There is no need to reverse the dimshuffle and reshape operations. I think what you are looking for is net['segLayer']. It has the exact shape you described in your comment: (batchsize, classes, height, width). I will edit the names so that they are more clear

Lasagne/Lasagne#626 alone would not help much because the shape after the new softmax is still incompatible with the categorical crossentropy implementation. 
I tried using
def softmax(x):
    e_x = theano.tensor.exp(x - x.max(axis=1, keepdims=True))
    return e_x / e_x.sum(axis=1, keepdims=True)
But the 4D output of the softmax did not work with the lasagne categorical crossentropy imlementation. I had a quick glance at the implementation: It did not look like changing it would be easy which is why I stuck with dimshuffle + reshape

f0k commented:
Hey Fabien, looks quite nice! I had a couple of suggestions for possible improvements, but it looks good in general.

> Is there any way to store the images as a .zip archive somewhere (~8GB)?

Is this the size of the compressed `.npy` arrays? We have some S3 storage, but I'm not sure if we want to host downloads that large -- so far we only use it for hosting pre-trained model weights. I can see that the download process is quite cumbersome, though.

ebenolson commented:
Thanks for this, it looks like a great example.

As @f0k said, we can store the data in S3 if necessary, but I'd like to bring the size down if possible. Compressed `.npy` may work, or maybe converting to `.jpg`. Would you get a decent result on a subset of the data? I guess probably not if you're doing data augmentation already...

Also, it seems like this would work well as an IPython notebook, many people would probably be satisfied viewing the results online and not need to download and train it themselves.

FabianIsensee commented:
Hi,
thank you for your feedback! So far the .npy files are not compressed, let me check tomorrow by how much I can reduce the file size.
Using less data would probably not work very well. Already now the results are worse than I would like them to be (96% accuracy sounds great but most of the image is background so that does not mean very much). As we speak my PC at work is running some more experiments in order to check if the results can get a bit better when using different hyperparameters and by tweaking some other stuff. It would be great if we could provide a model that works as well as possible so that people are motivated to dive into it a bit more!
Using an IPython notebook is a great idea (btw one of the reasons I started out in lasagne is because of your Neural networks with Theano and Lasagne talk on youtube ;-) ). However my IPython notebook does not find my Python Kernel. Since I myself mostly use plain IPython I did not put too much effort into fixing this. Let me have a look at it tomorrow. If I get it to run I will include a notebook as well. 
So far my pretrained weights are hosted on my personal dropbox. I don't mind the space it takes (its just some 25MB) but I am not sure how much traffic dropbox allows. Once we have a final version it would be great if you could host the pretrained weights.
Cheers,
Fabian

f0k commented on examples/UNet/massachusetts_road_segm.py:
```diff
@@ -52,12 +52,14 @@ def main():
-    target_valid= np.load("valid_target.npy", mmap_mode=mmap_mode)
-    data_test = np.load("test_data.npy", mmap_mode=mmap_mode)
-    target_test = np.load("test_target.npy", mmap_mode=mmap_mode)
+    dataset = np.load("road_segm_dataset.npz", mmap_mode=mmap_mode)
```
Memory-mapping only works for `.npy` files. The key idea is that the binary data of the `ndarray` is available as-is on disk, and can just be mapped into memory. This doesn't work when the binary data resides in a zip file.

How much smaller does the data get, though? What's the size of the `.npz` file?

FabianIsensee commented:
Uncompressed numpy arrays have a total size of 10.5GB, compressed npz file has 6.2GB. I guess that ist still too large for your S3 storage. I think the best solution is to let people download the image files themselves.
How many workers can I use do download the image files without having the server refuse the connection? I just downloaded the dataset with 4 workers and that took quite a long time, especially considering that we have >100Mbyte/s downstream speed...

FabianIsensee commented:
Most suggestions should be included by now. I have to leave for a family week end. Will continue working on everything on Monday.
Have a nice week-end!

f0k commented on examples/UNet/generators.py:
```diff
@@ -1,21 +1,30 @@
+        yield np.array(data[idx_2[ctr:ctr+BATCH_SIZE]]), np.array(target[idx_2[ctr:ctr+BATCH_SIZE]])
+        ctr += BATCH_SIZE
+        if ctr >= data.shape[0]:
+            ctr -= data.shape[0]
```
Sorry, this looks overly complicated. What about:
```python
def batch_generator(data, target, BATCH_SIZE, shuffle=False):
    if shuffle:
        while True:
            ids = np.random.choice(len(data))
            yield data[ids], target[ids]
    else:
        for idx in range(len(data) // BATCH_SIZE):
            ids = slice(idx * BATCH_SIZE, (idx + 1) * BATCH_SIZE)
            yield data[ids], target[ids]
```
This will discard the last incomplete batch. If you want to include it, you just need to change the range to get up to `np.ceil(len(data) / float(BATCH_SIZE))`.

FabianIsensee replied:
Thanks. I will use it! (btw you forgot to add BATCH_SIZE to this line: 
ids = np.random.choice(len(data))
)

At first I was not entirely convinced of this approach because this implementation behaves very differently depending on whether shuffle is True or False. True will yield an unlimited amount of training data (hence I need to manually break the train loop). False will break by itself (and I need to re-instantiate the generator every epoch)

f0k commented:
> How many workers can I use do download the image files without having the server refuse the connection?

I guess I'd just use 4 to be polite.

> I just downloaded the dataset with 4 workers and that took quite a long time

This shouldn't be a problem. If people know, they can let it download in the background and do something else. They'll also have to find something to do while training the network. But, out of curiosity: How long did it take?

Enjoy the weekend!

f0k commented on examples/UNet/generators.py:
```diff
@@ -21,6 +21,22 @@ def batch_generator(data, target, BATCH_SIZE, shuffle=False):
+    elif len(output_size) == 2:
+        center_crop = list(output_size)
+    else:
+        raise ValueError("invalid output_size")
```
Hint: You can use `center_crop = lasagne.utils.as_tuple(output_size, 2, int)`. This will take care of everything.

FabianIsensee replied:
Thanks. Will do!

The download took about an hour or maybe an hour and a half (using 4 threads)

benanne commented:
This looks great, thanks for contributing! I have one comment regarding evaluation: as you already pointed out measuring the accuracy can be a bit misleading in this setting. Due to the imbalance of the problem it might make more sense to report AUC scores or some form of precision / recall scores instead. When I used this dataset in a paper I reported the AUC, but the original work using this dataset (Vlad Mnih's PhD work) uses a more elaborate evaluation protocol.

FabianIsensee commented:
I think AUC will suffice for this educational example. It has been added with the last commit

FabianIsensee commented:
Alright I think all of your suggestions are now incorporated into the pull request. If anything else needs changing please just let me know! 
I added a new set of pretrained weights for the net:
# val accuracy:  0.963513  val loss:  0.114994  val AUC score:  0.978996643458
# https://www.dropbox.com/s/0vasqq491skf9iz/UNet_mass_road_segm_params.zip?dl=0
The .zip archive is only 9 MB large. I will leave it up to you whether you want to host it yourself or just leave it on my Dropbox. The dataset is simply too large to host it. Users will have to download it themselves. 
Thank you very much for your constructive criticism so far! It really helped not just to make this pull request more mature but also in teaching me a lot of new things!

FabianIsensee commented:
Hi,
what is the status of the pull request? What do you think about providing both implementations?
I can understand if this fits more in the papers section. If you want I can rearrange my code accordingly.
Cheers,
Fabian

ebenolson commented:
Hey, sorry for the late response. I'm happy to merge as is or if you want to reformat for `/papers/` that's fine too.

I uploaded the weights to https://s3.amazonaws.com/lasagne/recipes/pretrained/UNet_mass_road_segm_params.zip, if you want to change it to point to there. I think you're right that the dataset it too large to rehost, we can revisit that if the original goes down for some reason.

Thanks again for your contribution.

ebenolson commented:
Hi @FabianIsensee - I'd like to get this merged, if you have a chance could you update the url for the weights and squash your history to a single commit? If you're too busy I'll do it, just need to figure out how to retain your authorship.

FabianIsensee commented:
Hi,
sorry for keeping you waiting this long. I will do it within the next days and let you know once everything is completed.
Cheers,
Fabian

FabianIsensee commented:
Hi,
I am almost finished with the PR.
My original implementation of UNet was not very close to the paper. I adapted it to be almost identical (except for pad 'same' instead of valid and my use of batch_norm). This adaptation requires new pretrained weights. It would be great if you could update the weights on aws with these: https://www.dropbox.com/s/i9sara0ds90d79b/pretrained_weights.zip?dl=0
I will post again once my commits are squashed.
Cheers,
Fabian

FabianIsensee commented:
Mhm I am not much of a git ninja. I just attempted to squash my commits but am unsure whether this actually worked. This is the commit where I squashed everything: d25461731732487ae8a6dc9ea6c338ef5986018b
Please let me know whether this is wat you need ;-) If so then the PR is ready to go.
(Please remember to update the weights on aws)
Cheers,
Fabian

ebenolson commented:
Hmm, apparently github now lets me squash directly from the merge button... I've updated the weights and I'll merge now.

Thank you again for the great example!

FabianIsensee commented:
Alright then! Glad you like it!

steven2358 commented:
@FabianIsensee You mentioned in a code comment that the paper does not describe the dropout configuration.

This can be found in the authors' implementation. In particular, one dropout layer with p=0.5 is added after `net['contr_3_2']` and another one after `net['contr_4_2']`.

Source: http://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/
